### ISSUE TITLE
TypeError: Path must be a string. Received undefined

### ISSUE THREAD

**Issue by simonhaenisch** [browserify/resolve#181]
Following up on https://github.com/browserify/resolve/issues/177#issuecomment-453364771 (not sure whether it's related), `1.9.0` has this code:

https://github.com/browserify/resolve/blob/254bb4029df2f8d20a33043dfabd8e5cabfa37df/lib/async.js#L47-L50

which throws an error if `fs.realpath` fails and the error is `ENOENT`, because then `realStart` is not defined.

```
TypeError: Path must be a string. Received undefined
```

This was already fixed in master in a commit prior to the release `1.9.0` (https://github.com/browserify/resolve/commit/2321cd4b4e184f9c97f76ea50dd49b9e42927b93#diff-688ede0ee2da1fa3d89f2fc485f39273), however has not yet been released.

Here is the same part with the fix on master:

https://github.com/browserify/resolve/blob/7c6afabf856bc0c4a268d1ba0d3328abef4ba8eb/lib/async.js#L58-L61

Can we get this fix out any time soon?

**Comment by simonhaenisch** [comment]
It's not in the `v1.9.0` tag that was released to npm... see

https://github.com/browserify/resolve/blob/v1.9.0/lib/async.js#L49

**Comment by ljharb** [comment]
The `realpath` fix is; the "throw on invalid directory" breaking change is not (and won't be).

**Comment by ljharb** [comment]
@simonhaenisch i'm super on board with releasing a v1.9.1 that fixes this for you; is there any chance you could make a PR into v1.9.0 (not master) with a failing test case?

**Comment by simonhaenisch** [comment]
Please see https://github.com/browserify/resolve/compare/v1.9.0...simonhaenisch:fix-path-must-be-string, here is the test results without and with the fix.

**Before:**

```
TAP version 13
# non-existent basedir should not throw when preserveSymlinks is false
path.js:28
    throw new TypeError('Path must be a string. Received ' + inspect(path));
    ^

TypeError: Path must be a string. Received undefined
    at assertPath (path.js:28:11)
    at Object.resolve (path.js:1168:7)
    at init (~/resolve/lib/async.js:58:24)
    at ~/resolve/lib/async.js:49:18
    at gotStat (fs.js:1783:21)
    at FSReqWrap.oncomplete (fs.js:152:21)
```

**After:**

```
# non-existent basedir should not throw when preserveSymlinks is false
ok 1 should be equal
ok 2 should be equal

1..2
# tests 2
# pass  2

# ok
```

_(ran with Node v8.15.0)_

---

My branch is based of `v1.9.0`, but I'm not sure how to open a PR against that because it seems I can only open one against a branch but not a tag. Maybe you'll have to `git checkout -b v1.9.0-patch v1.9.0` first and push it, so I can open the PR?

I don't have a Windows instance to run the tests against btw.

**Comment by simonhaenisch** [comment]
@ljharb could you please create a branch based of the `v1.9.0` tag and push it to remote, so I can open the PR?

**Comment by ljharb** [comment]
@simonhaenisch sorry for the delay; done! https://github.com/browserify/resolve/tree/1.9.0-next

**Comment by simonhaenisch** [comment]
Thanks! PR is open (:

**Comment by ljharb** [comment]
Merged into the next v1 release (will be v1.10.0); not yet merged into master. When it is, this will close.

**Comment by ljharb** [comment]
v1.10.0 is released; merged into master with ea7d38d8212b3c1d2e599812008b3670241ad12f
